File: ios/ZapLivre/ZapLivre/Core/MediaPickerViewModel.swift
//
//  MediaPickerViewModel.swift
//  ZapLivre
//
//  ViewModel for managing media selection and upload
//

import SwiftUI
import Combine

/// Upload state
enum UploadState: Equatable {
    case idle
    case uploading(current: Int, total: Int)
    case success
    case error(String)
}

/// Media picker view model
class MediaPickerViewModel: ObservableObject {
    @Published var selectedImages: [UIImage] = []
    @Published var uploadState: UploadState = .idle

    private let zapLivreCore: ZapLivreCore

    init(zapLivreCore: ZapLivreCore = .shared) {
        self.zapLivreCore = zapLivreCore
    }

    /// Add images to selection
    func addImages(_ images: [UIImage]) {
        selectedImages.append(contentsOf: images)
    }

    /// Remove image at index
    func removeImage(at index: Int) {
        guard index >= 0 && index < selectedImages.count else { return }
        selectedImages.remove(at: index)
    }

    /// Clear all selected images
    func clearSelection() {
        selectedImages.removeAll()
    }

    /// Upload images to peer
    func uploadImages(to peerId: String, quality: CGFloat = 0.85) {
        guard !selectedImages.isEmpty else { return }

        uploadState = .uploading(current: 0, total: selectedImages.count)

        Task {
            do {
                for (index, image) in selectedImages.enumerated() {
                    try await uploadSingleImage(image, to: peerId, quality: quality)

                    await MainActor.run {
                        uploadState = .uploading(current: index + 1, total: selectedImages.count)
                    }
                }

                await MainActor.run {
                    uploadState = .success
                    clearSelection()
                }
            } catch {
                await MainActor.run {
                    uploadState = .error(error.localizedDescription)
                }
            }
        }
    }

    /// Upload a single image with compression (via FFI)
    private func uploadSingleImage(_ image: UIImage, to peerId: String, quality: CGFloat) async throws {
        // Convert UIImage to JPEG data (pre-compression before FFI)
        guard let imageData = image.jpegData(compressionQuality: quality) else {
            throw MediaError.compressionFailed
        }

        // Generate unique filename
        let fileName = "image_\(Int(Date().timeIntervalSince1970)).jpg"

        // Convert quality to 0-100 scale for FFI
        let qualityPercent = UInt32(quality * 100)

        // Call FFI method to send image with additional compression in Rust
        let messageId = try await zapLivreCore.sendImageMessage(
            to: peerId,
            imageData: imageData,
            fileName: fileName,
            quality: qualityPercent
        )

        print("✅ Image sent successfully: \(messageId)")
    }

    /// Reset upload state
    func resetUploadState() {
        uploadState = .idle
    }

    /// Get compressed JPEG data from UIImage
    func getCompressedJPEG(from image: UIImage, quality: CGFloat = 0.8) -> Data? {
        return image.jpegData(compressionQuality: quality)
    }

    /// Get thumbnail from UIImage
    func getThumbnail(from image: UIImage, size: CGSize = CGSize(width: 200, height: 200)) -> UIImage? {
        let renderer = UIGraphicsImageRenderer(size: size)
        return renderer.image { _ in
            image.draw(in: CGRect(origin: .zero, size: size))
        }
    }
}

/// Media-related errors
enum MediaError: LocalizedError {
    case compressionFailed
    case uploadFailed
    case invalidImage

    var errorDescription: String? {
        switch self {
        case .compressionFailed:
            return "Failed to compress image"
        case .uploadFailed:
            return "Failed to upload image"
        case .invalidImage:
            return "Invalid image format"
        }
    }
}

/// Media item metadata
struct MediaItem: Identifiable {
    let id = UUID()
    let image: UIImage
    let fileName: String?
    let fileSize: Int64?
}

File: android/app/src/main/kotlin/com/zaplivre/core/MediaPickerViewModel.kt
package com.zaplivre.core

import android.content.Context
import android.net.Uri
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import uniffi.zaplivre.ZapLivreClient

/**
 * ViewModel for managing media (images) selection and upload
 */
class MediaPickerViewModel(
    private val client: ZapLivreClient,
    private val context: Context
) : ViewModel() {

    private val _selectedImages = MutableStateFlow<List<MediaItem>>(emptyList())
    val selectedImages: StateFlow<List<MediaItem>> = _selectedImages.asStateFlow()

    private val _uploadState = MutableStateFlow<UploadState>(UploadState.Idle)
    val uploadState: StateFlow<UploadState> = _uploadState.asStateFlow()

    /**
     * Add images from URIs (from Photo Picker)
     */
    fun addImages(uris: List<Uri>) {
        viewModelScope.launch {
            val newItems = uris.map { uri ->
                MediaItem(
                    uri = uri,
                    type = MediaType.IMAGE,
                    fileName = getFileName(uri),
                    fileSize = getFileSize(uri)
                )
            }
            _selectedImages.value = _selectedImages.value + newItems
        }
    }

    /**
     * Remove an image from selection
     */
    fun removeImage(uri: Uri) {
        _selectedImages.value = _selectedImages.value.filterNot { it.uri == uri }
    }

    /**
     * Clear all selected images
     */
    fun clearSelection() {
        _selectedImages.value = emptyList()
    }

    /**
     * Upload selected images to a peer
     */
    fun uploadImages(toPeerId: String, quality: UInt = 85u) {
        if (_selectedImages.value.isEmpty()) return

        viewModelScope.launch {
            _uploadState.value = UploadState.Uploading(0, _selectedImages.value.size)

            try {
                _selectedImages.value.forEachIndexed { index, mediaItem ->
                    uploadSingleImage(toPeerId, mediaItem, quality)

                    _uploadState.value = UploadState.Uploading(
                        current = index + 1,
                        total = _selectedImages.value.size
                    )
                }

                _uploadState.value = UploadState.Success
                clearSelection()
            } catch (e: Exception) {
                _uploadState.value = UploadState.Error(e.message ?: "Upload failed")
            }
        }
    }

    /**
     * Upload a single image (with compression via FFI)
     */
    private suspend fun uploadSingleImage(
        toPeerId: String,
        mediaItem: MediaItem,
        quality: UInt
    ) {
        withContext(Dispatchers.IO) {
            // Read image bytes from URI
            val inputStream = context.contentResolver.openInputStream(mediaItem.uri)
                ?: throw Exception("Failed to open image: ${mediaItem.uri}")

            val imageBytes = inputStream.use { it.readBytes() }

            // Call FFI method to send image with compression
            // The compression happens in the Rust core via compress_image()
            val messageId = client.sendImageMessage(
                toPeerId = toPeerId,
                imageData = imageBytes.toUByteArray().toList(),
                fileName = mediaItem.fileName ?: "image_${System.currentTimeMillis()}.jpg",
                quality = quality
            )

            // Message sent successfully, messageId returned
            // UI will be updated via message events from the core
        }
    }

    /**
     * Get file name from URI
     */
    private fun getFileName(uri: Uri): String? {
        return context.contentResolver.query(uri, null, null, null, null)?.use { cursor ->
            val nameIndex = cursor.getColumnIndex(android.provider.OpenableColumns.DISPLAY_NAME)
            if (cursor.moveToFirst() && nameIndex != -1) {
                cursor.getString(nameIndex)
            } else {
                null
            }
        }
    }

    /**
     * Get file size from URI
     */
    private fun getFileSize(uri: Uri): Long? {
        return context.contentResolver.query(uri, null, null, null, null)?.use { cursor ->
            val sizeIndex = cursor.getColumnIndex(android.provider.OpenableColumns.SIZE)
            if (cursor.moveToFirst() && sizeIndex != -1) {
                cursor.getLong(sizeIndex)
            } else {
                null
            }
        }
    }

    /**
     * Reset upload state
     */
    fun resetUploadState() {
        _uploadState.value = UploadState.Idle
    }
}

/**
 * Media item (image, video, file)
 */
data class MediaItem(
    val uri: Uri,
    val type: MediaType,
    val fileName: String?,
    val fileSize: Long?
)

/**
 * Media type
 */
enum class MediaType {
    IMAGE,
    VIDEO,
    DOCUMENT,
    AUDIO
}

/**
 * Upload state
 */
sealed class UploadState {
    object Idle : UploadState()
    data class Uploading(val current: Int, val total: Int) : UploadState()
    object Success : UploadState()
    data class Error(val message: String) : UploadState()
}
